07_VirtualFileSystem/02_VirtualFileSystem.md

```diff
@@ -0,0 +1,352 @@
+
+```c
+struct {
+    int fs_file_id;
+    int mountpoint_id;
+    char *filename;
+    int buf_read_pos;
+    int buf_write_pos;
+    char *file_buffer;
+} file_descriptor_t
+```
+
+Where the `mountpoint_id` fields is the id of the mounted file system that is contining the requeste file. The `fs_file_id` is the fs specific id of the fs opened file descriptor. `buf_read_pos` and `buf_write_pos` are the current positions of the buffer pointer for the read and write operations. 
+
+So once our open function has found the mountpoint for the requested file, eventually a new file descriptor item will be created and filled, and an id value returned. This id is different from the ine in the data structure, since it represent the internal fs descriptor id, while this one represent the vfs descriptor id. In our case the descriptor list is implemented again using an array, so the id returned will be the array position where the descriptor is being filled.
+
+Why "eventually" ? Having found the mountpoint id for the file doesn't mean that the file exists on that fs, the only thing that exist so far is the mountpoint, but after that the VFS can't really know if the file exists or not, it has to devolve this task to the fs driver, hence it will call the implementation of a function that open a file on that FS that will do the search and return the an error if the file doesn't exists.  
+
+But how to call the fs driver function? Earlier in this chapter when we outlined the `mountpoint_t` structure we added a field called `operations`, of type `fs_operations_t` and left it unimplemented. Now is the time to implement it, this field is going to contain the pointer to the driver functions that will be used by the vfs to open, read, write, and close files:
+
+```c
+struct fs_operations_t {
+	int (*open)(const char *path, int flags, ... );
+	int (*close)(int);
+	ssize_t (*read)(int, char*, size_t);
+	ssize_t (*write)(int,const void*, size_t);
+};
+
+typedef struct fs_operations_t fs_operations_t;
+```
+
+The basic idea is that once the mountpoint_id has been found (we declared `get_mountpoint_id` in the previous paragraph), the vfs will use the mountpoint item to call the fs driver implementation of the open function, the only thing we need to consider is that when calling the driver function, it doesn't care about the mountpoint part of the path, and if the whole path is passed, we will most likely get an error, since the fs root will start from within the mountpoint folder. What we need thean is a function that get the relative path from an absolute one, the idea and implementation should be pretty straightforward, and just involve few `strcpy` operations, so this is left as exercise, let's just assume that we have one with the following signature: 
+
+```c
+char *get_rel_path(char *mountpoint_part, char* full_path);
```

Comment:
This relative path stuff should probably go with the other stuff on relative paths (a few paragraphs earlier).